cli/cmd/cluster_shell.go

```diff
@@ -0,0 +1,141 @@
+	cmd.Flags().StringVar(&r.args.shellClusterID, "id", "", "id of the cluster to have kubectl access to (when name is not provided)")
+
+	return cmd
+}
+
+func (r *runners) shellCluster(_ *cobra.Command, args []string) error {
+	// by default, we look at args[0] as the id
+	// but if it's not provided, we look for a viper flag named "name" and use it
+	// as the name of the cluster, not the id
+	clusterID := ""
+	if len(args) > 0 {
+		clusterID = args[0]
+	} else if r.args.shellClusterName != "" {
+		clusters, err := r.kotsAPI.ListClusters(false, nil, nil)
+		if errors.Cause(err) == platformclient.ErrForbidden {
+			return ErrCompatibilityMatrixTermsNotAccepted
+		} else if err != nil {
+			return errors.Wrap(err, "list clusters")
+		}
+		for _, cluster := range clusters {
+			if cluster.Name == r.args.shellClusterName {
+				clusterID = cluster.ID
+				break
+			}
+		}
+	} else if r.args.shellClusterID != "" {
+		clusterID = r.args.shellClusterID
+	} else {
+		return errors.New("must provide cluster id or name")
+	}
+
+	cluster, err := r.kotsAPI.GetCluster(clusterID)
+	if errors.Cause(err) == platformclient.ErrForbidden {
+		return ErrCompatibilityMatrixTermsNotAccepted
+	} else if err != nil {
+		return errors.Wrap(err, "get cluster")
+	}
+
+	if cluster.Status != types.ClusterStatusRunning {
+		return errors.Errorf("cluster %s is not %s, please check the cluster status", clusterID, types.ClusterStatusRunning)
+	}
+
+	kubeconfig, err := r.kotsAPI.GetClusterKubeconfig(clusterID)
+	if errors.Cause(err) == platformclient.ErrForbidden {
+		return ErrCompatibilityMatrixTermsNotAccepted
+	} else if err != nil {
+		return errors.Wrap(err, "get cluster kubeconfig")
+	}
+
+	tmpFile, err := os.CreateTemp("", "replicated-kubeconfig")
+	if err != nil {
+		return errors.Wrap(err, "create temp file")
+	}
+	defer func() {
+		tmpFile.Close()
+		os.Remove(tmpFile.Name())
+	}()
+	if _, err := tmpFile.Write(kubeconfig); err != nil {
+		return errors.Wrap(err, "write kubeconfig file")
+	}
+	if err := tmpFile.Chmod(0644); err != nil {
+		return errors.Wrap(err, "chmod kubeconfig file")
+	}
+
+	shellCmd := os.Getenv("SHELL")
+	if shellCmd == "" {
+		return errors.New("SHELL environment is required for shell command")
+	}
+
+	shellExec := exec.Command(shellCmd)
+	shellExec.Env = os.Environ()
+	fmt.Printf("Starting new shell with KUBECONFIG. Press Ctl-D when done to end the shell and the connection to the server\n")
+	shellPty, err := pty.Start(shellExec)
+	if err != nil {
+		return errors.Wrap(err, "failed to start shell")
+	}
+
+	// Handle pty size.
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, syscall.SIGWINCH)
+	go func() {
+		for range ch {
+			if err := pty.InheritSize(os.Stdin, shellPty); err != nil {
+				log.Printf("error resizing pty: %s", err)
+			}
+		}
+	}()
+	ch <- syscall.SIGWINCH // Initial resize.
+	defer func() { signal.Stop(ch); close(ch) }()
+
+	// Set stdin to raw mode.
+	oldState, err := term.MakeRaw(int(os.Stdin.Fd()))
```

Comment:
ive never worked with this library. what happens if i hit control c? how do i restore the terminal?

Reply:
control c won't do anything. control d will exit.
This is the same lib as used by [embedded cluster](https://github.com/replicatedhq/embedded-cluster/blob/main/cmd/embedded-cluster/shell.go#L61) and [sbctl](https://github.com/replicatedhq/sbctl/blob/main/cli/shell.go#L141)